# Omar Hamoudeh
# Programming Exercise \#1
---

# Preamble

In [3]:
import numpy as np
import pandas as pd
from scipy import stats as sps
from IPython.display import display, Latex
pd.set_option('display.max_columns', None)


# **1. Fetal Health Classification Dataset**

## **Clean Dataset**

### Problem 1.1

In [4]:
# load the clean dataset csv file into a pandas dataframe

fetal_df = pd.read_csv('fetal_health_dataset_clean.csv')


#### (a)

This is a supervised machine learning task because the dataset contains a labeled column, 'fetal_health,' which categorizes each data point (fetus) into one of three classes: Normal, Suspect, or Pathological. Since the goal is to predict the health category of the fetus based on other features in the dataset, and the labels are already provided for the model to learn from, it fits the definition of a supervised learning problem.

#### (b)

In [5]:
### Your code for 1.1(b) goes here ###
print("Axes:", fetal_df.axes)
print("-----------------------------------------------------------------------------------")
print("Data types:")
display(fetal_df.dtypes)

Axes: [RangeIndex(start=0, stop=2126, step=1), Index(['baseline value', 'accelerations', 'fetal_movement',
       'uterine_contractions', 'light_decelerations', 'severe_decelerations',
       'prolongued_decelerations', 'abnormal_short_term_variability',
       'mean_value_of_short_term_variability',
       'percentage_of_time_with_abnormal_long_term_variability',
       'mean_value_of_long_term_variability', 'histogram_width',
       'histogram_min', 'histogram_max', 'histogram_number_of_peaks',
       'histogram_number_of_zeroes', 'histogram_mode', 'histogram_mean',
       'histogram_median', 'histogram_variance', 'histogram_tendency',
       'fetal_health'],
      dtype='object')]
-----------------------------------------------------------------------------------
Data types:


baseline value                                              int64
accelerations                                             float64
fetal_movement                                            float64
uterine_contractions                                      float64
light_decelerations                                       float64
severe_decelerations                                      float64
prolongued_decelerations                                  float64
abnormal_short_term_variability                             int64
mean_value_of_short_term_variability                      float64
percentage_of_time_with_abnormal_long_term_variability      int64
mean_value_of_long_term_variability                       float64
histogram_width                                             int64
histogram_min                                               int64
histogram_max                                               int64
histogram_number_of_peaks                                   int64
histogram_

#### (c)



In [6]:
display(fetal_df.head(10))

baseline value  accelerations  fetal_movement  uterine_contractions  \
0             120          0.000             0.0                 0.000   
1             132          0.006             0.0                 0.006   
2             133          0.003             0.0                 0.008   
3             134          0.003             0.0                 0.008   
4             132          0.007             0.0                 0.008   
5             134          0.001             0.0                 0.010   
6             134          0.001             0.0                 0.013   
7             122          0.000             0.0                 0.000   
8             122          0.000             0.0                 0.002   
9             122          0.000             0.0                 0.003   

   light_decelerations  severe_decelerations  prolongued_decelerations  \
0                0.000                   0.0                     0.000   
1                0.003                   0.0                     0.000   
2                0.003                   0.0                     0.000   
3                0.003                   0.0                     0.000   
4                0.000                   0.0                     0.000   
5                0.009                   0.0                     0.002   
6                0.008                   0.0                     0.003   
7                0.000                   0.0                     0.000   
8                0.000                   0.0                     0.000   
9                0.000                   0.0                     0.000   

   abnormal_short_term_variability  mean_value_of_short_term_variability  \
0                               73                                   0.5   
1                               17                                   2.1   
2                               16                                   2.1   
3                               16                                   2.4   
4                               16                                   2.4   
5                               26                                   5.9   
6                               29                                   6.3   
7                               83                                   0.5   
8                               84                                   0.5   
9                               86                                   0.3   

   percentage_of_time_with_abnormal_long_term_variability  \
0                                                 43        
1                                                  0        
2                                                  0        
3                                                  0        
4                                                  0        
5                                                  0        
6                                                  0        
7                                                  6        
8                                                  5        
9                                                  6        

   mean_value_of_long_term_variability  histogram_width  histogram_min  \
0                                  2.4               64             62   
1                                 10.4              130             68   
2                                 13.4              130             68   
3                                 23.0              117             53   
4                                 19.9              117             53   
5                                  0.0              150             50   
6                                  0.0              150             50   
7                                 15.6               68             62   
8                                 13.6               68             62   
9                                 10.6               68             62   

   histogram_max  histogram_number_of_peaks  hist

#### (d)

In [7]:
### Your code for 1.1(d) goes here ###
print("Rows:", fetal_df.shape[0])
print("Columns:", fetal_df.shape[1])


#for col in fetal_df.columns:
#    unique_vals = pd.unique(fetal_df[col])
#    if len(unique_vals) < 9:
#        print(f"Column: {col}, Unique values: {len(unique_vals)} Values: {unique_vals}")


Rows: 2126
Columns: 22


#### (e)

Each row of fetal_df dataframe can be termed as a datapoint, sample, or observation.

#### (f)

There are 2126 samples in this dataset 

#### (g)

There are a total of 21 Independent variables in this dataset. They are listed below:

baseline value  
accelerations  
fetal_movement  
uterine_contractions  
light_decelerations  
severe_decelerations  
prolongued_decelerations  
abnormal_short_term_variability  
mean_value_of_short_term_variability  
percentage_of_time_with_abnormal_long_term_variability  
mean_value_of_long_term_variability  
histogram_width  
histogram_min  
histogram_max  
histogram_number_of_peaks  
histogram_number_of_zeroes  
histogram_mode  
histogram_mean  
histogram_median  
histogram_variance  
histogram_tendency  


#### (h)

There is 1 Dependent variable in this data set which is 'fetal_health'

#### (i)

n = 2126

p = 21

#### (j)

n = 2126

p = 1

#### (k)

I believe that most of the data in the dataset has been pre-processed into transformed features. While there are a few variables that may represent raw sensor data, such as 'baseline value' and 'fetal_movement', the majority of the columns suggest that the data has undergone some form of pre-processing. For example, the various histogram features (like min, max, median, mode, mean, and variance) are not raw sensor outputs but derived features that must be extracted through statistical calculations based on the sensor data.

#### (l)

There are a total of 2 categorical variables in the dataset which are 'histogram_tendency' and 'fetal_health'.  
I figured this out by using the pandas.unique() function and filtered for variables with less than 10 unique values.  
'histogram_tendency' only had 3 unique values which were {-1, 0, 1}, which refered to negative, neutral, and positive tendencies respectivley which are ordinal categories.  
'fetal_health' also only had 3 unique values which were {1, 2, 3}, which refered to normal, suspect, and pathological which are nominal categories.

#### (m)

histogram_tendency: Ordinal encoding {-1, 0, 1}  
fetal_health: Nominal encoding {1, 2, 3}

#### (n)

In [8]:
normal = 0
suspect = 0
pathological = 0
error = 0

for sample in fetal_df.itertuples():
    if sample.fetal_health == 1: normal += 1
    elif sample.fetal_health == 2: suspect += 1
    elif sample.fetal_health == 3: pathological += 1
    else: error += 1

print(f"There are {normal} samples with normal health condition.")
print(f"There are {suspect} samples with suspect health condition.")
print(f"There are {pathological} samples with pathological health condition.")

There are 1655 samples with normal health condition.
There are 295 samples with suspect health condition.
There are 176 samples with pathological health condition.


## **Dirty Dataset**

In [9]:
# load the dirty dataset csv file into a pandas dataframe
fetal_dirty_df = pd.read_csv('fetal_health_dataset_dirty.csv')

### Problem 1.2

#### (a)

In [10]:
# Outputs a series labeling columns with NaN values True, and False otherwise
NaN_by_column = fetal_dirty_df.isna().any()
# Checks if there is a True in the Series
print(NaN_by_column.any())

True


#### (b)

In [11]:

num_of_missing_values = fetal_dirty_df.isna().sum()
cols_missing_values = num_of_missing_values[num_of_missing_values > 0]
display(cols_missing_values)


baseline value                                            47
percentage_of_time_with_abnormal_long_term_variability    10
histogram_max                                             24
histogram_mode                                             5
dtype: int64

#### (c)

In [12]:
print("Number of samples with missing values: ", fetal_dirty_df.isna().any(axis=1).sum())

Number of samples with missing values:  86


### Problem 1.3

In [13]:

fetal_dirty_df['baseline value'] = np.where(
    (fetal_dirty_df['baseline value'] < 110) | 
    (fetal_dirty_df['baseline value'] > 160), 
    np.nan, fetal_dirty_df['baseline value'])

fetal_dirty_df['accelerations'] = np.where(
    (fetal_dirty_df['accelerations'] < 0) | 
    (fetal_dirty_df['accelerations'] > 0.02), 
    np.nan, fetal_dirty_df['accelerations'])

fetal_dirty_df['fetal_movement'] = np.where(
    (fetal_dirty_df['fetal_movement'] < 0.001) | 
    (fetal_dirty_df['fetal_movement'] > 0.005), 
    np.nan, fetal_dirty_df['fetal_movement'])

fetal_dirty_df['uterine_contractions'] = np.where(
    (fetal_dirty_df['uterine_contractions'] < 0) | 
    (fetal_dirty_df['uterine_contractions'] > 0.01), 
    np.nan, fetal_dirty_df['uterine_contractions'])

fetal_dirty_df['light_decelerations'] = np.where(
    (fetal_dirty_df['light_decelerations'] < 0) | 
    (fetal_dirty_df['light_decelerations'] > 0.01), 
    np.nan, fetal_dirty_df['light_decelerations'])

fetal_dirty_df['severe_decelerations'] = np.where(
    (fetal_dirty_df['severe_decelerations'] < 0) | 
    (fetal_dirty_df['severe_decelerations'] > 0.01), 
    np.nan, fetal_dirty_df['severe_decelerations'])

fetal_dirty_df['prolongued_decelerations'] = np.where(
    (fetal_dirty_df['prolongued_decelerations'] < 0) | 
    (fetal_dirty_df['prolongued_decelerations'] > 0.01), 
    np.nan, fetal_dirty_df['prolongued_decelerations'])

fetal_dirty_df['abnormal_short_term_variability'] = np.where(
    (fetal_dirty_df['abnormal_short_term_variability'] < 10) | 
    (fetal_dirty_df['abnormal_short_term_variability'] > 90), 
    np.nan, fetal_dirty_df['abnormal_short_term_variability'])

fetal_dirty_df['mean_value_of_short_term_variability'] = np.where(
    (fetal_dirty_df['mean_value_of_short_term_variability'] < 0.2) | 
    (fetal_dirty_df['mean_value_of_short_term_variability'] > 5), 
    np.nan, fetal_dirty_df['mean_value_of_short_term_variability'])

fetal_dirty_df['percentage_of_time_with_abnormal_long_term_variability'] = np.where(
    (fetal_dirty_df['percentage_of_time_with_abnormal_long_term_variability'] < 0) | 
    (fetal_dirty_df['percentage_of_time_with_abnormal_long_term_variability'] > 80), 
    np.nan, fetal_dirty_df['percentage_of_time_with_abnormal_long_term_variability'])

fetal_dirty_df['mean_value_of_long_term_variability'] = np.where(
    (fetal_dirty_df['mean_value_of_long_term_variability'] < 0) | 
    (fetal_dirty_df['mean_value_of_long_term_variability'] > 35), 
    np.nan, fetal_dirty_df['mean_value_of_long_term_variability'])

fetal_dirty_df['histogram_width'] = np.where(
    fetal_dirty_df['histogram_width'] < 0, 
    np.nan, fetal_dirty_df['histogram_width'])

fetal_dirty_df['histogram_min'] = np.where(
    fetal_dirty_df['histogram_min'] < 50, 
    np.nan, fetal_dirty_df['histogram_min'])

fetal_dirty_df['histogram_max'] = np.where(
    fetal_dirty_df['histogram_max'] < 0, 
    np.nan, fetal_dirty_df['histogram_max'])

fetal_dirty_df['histogram_number_of_peaks'] = np.where(
    (fetal_dirty_df['histogram_number_of_peaks'] < 0) | 
    (fetal_dirty_df['histogram_number_of_peaks'] % 1 != 0), 
    np.nan, fetal_dirty_df['histogram_number_of_peaks'])

fetal_dirty_df['histogram_number_of_zeroes'] = np.where(
    (fetal_dirty_df['histogram_number_of_zeroes'] < 0) | 
    (fetal_dirty_df['histogram_number_of_zeroes'] % 1 != 0), 
    np.nan, fetal_dirty_df['histogram_number_of_zeroes'])

fetal_dirty_df['histogram_mode'] = np.where(
    fetal_dirty_df['histogram_mode'] < 0, 
    np.nan, fetal_dirty_df['histogram_mode'])

fetal_dirty_df['histogram_mean'] = np.where(
    fetal_dirty_df['histogram_mean'] < 0, 
    np.nan, fetal_dirty_df['histogram_mean'])

fetal_dirty_df['histogram_median'] = np.where(
    fetal_dirty_df['histogram_median'] < 0, 
    np.nan, fetal_dirty_df['histogram_median'])

fetal_dirty_df['histogram_variance'] = np.where(
    fetal_dirty_df['histogram_variance'] < 0, 
    np.nan, fetal_dirty_df['histogram_variance'])

fetal_dirty_df['histogram_tendency'] = np.where(
    ~fetal_dirty_df['histogram_tendency'].isin([-1, 0, 1]), 
    np.nan, fetal_dirty_df['histogram_tendency'])




### Problem 1.4

In [14]:

def zero_mean_and_unit_variance(df, cols_to_exclude):
    df.loc[:, df.columns.difference(cols_to_exclude)] = df.loc[:, df.columns.difference(cols_to_exclude)].apply(lambda x: (x - x.mean()) / x.std())


categorical_cols = ['histogram_tendency','fetal_health']
zero_mean_and_unit_variance(fetal_dirty_df, categorical_cols)

display(fetal_dirty_df.head(20))



baseline value  accelerations  fetal_movement  uterine_contractions  \
0        -1.375981      -0.822195             NaN             -1.509123   
1        -0.142088       0.729961             NaN              0.619849   
2        -0.039264      -0.046117             NaN              1.329506   
3         0.063561      -0.046117             NaN              1.329506   
4        -0.142088       0.988654             NaN              1.329506   
5         0.063561      -0.563502             NaN              2.039164   
6         0.063561      -0.563502             NaN                   NaN   
7        -1.170332      -0.822195             NaN             -1.509123   
8        -1.170332      -0.822195             NaN             -0.799466   
9        -1.170332      -0.822195             NaN             -0.444637   
10        1.811576      -0.822195             NaN             -1.154295   
11        1.708751      -0.822195             NaN             -1.154295   
12       -0.244913       0.471268             NaN              1.329506   
13       -0.244913       1.506039             NaN              0.619849   
14       -0.347737       0.729961             NaN             -0.089809   
15       -0.347737       0.729961             NaN             -0.089809   
16       -0.347737       0.729961             NaN              0.265020   
17       -0.244913      -0.304810             NaN             -0.444637   
18       -0.347737      -0.046117             NaN              0.619849   
19       -0.347737       0.471268             NaN              0.265020   

    light_decelerations  severe_decelerations  prolongued_decelerations  \
0             -0.652153             -0.057462                 -0.268691   
1              0.517903             -0.057462                 -0.268691   
2              0.517903             -0.057462                 -0.268691   
3              0.517903             -0.057462                 -0.268691   
4             -0.652153             -0.057462                 -0.268691   
5              2.858015             -0.057462                  3.121441   
6              2.467996             -0.057462                  4.816507   
7             -0.652153             -0.057462                 -0.268691   
8             -0.652153             -0.057462                 -0.268691   
9             -0.652153             -0.057462                 -0.268691   
10            -0.262135             -0.057462                 -0.268691   
11            -0.262135             -0.057462                 -0.268691   
12             0.517903             -0.057462                 -0.268691   
13             0.127884             -0.057462                 -0.268691   
14             1.297940             -0.057462                  1.426375   
15             0.907922             -0.057462                  1.426375   
16             1.297940             -0.057462                 -0.268691   
17             1.297940             -0.057462                  3.121441   
18             0.907922             -0.057462                  1.426375   
19             0.907922             -0.057462                  1.426375   

    abnormal_short_term_variability  mean_value_of_short_term_variability  \
0                          1.512834                             -0.987334   
1                         -1.744341                              0.965863   
2                         -1.802504                              0.965863   
3                         -1.802504                              1.332087   
4                         -1.802504                              1.332087   
5                         -1.220866                                   NaN   
6                         -1.046375                                   NaN   
7                          2.094473                             -0.987334   
8                          2.152636                             -0.987334   
9                          2.268964                             -1.231484   
10   

### Problem 1.5

In [15]:
fetal_dirty_df = fetal_dirty_df.groupby('fetal_health').apply(lambda group: group.fillna(group.median()))

### Problem 1.6

In [16]:
one_hot_encoded = pd.get_dummies(fetal_dirty_df, columns=['fetal_health'], prefix='fetal_health')

one_hot_encoded.to_csv('fetal_heatlh_dataset_processed.csv', index=False)

# **2. Heart Failure Prediction Dataset**

## Problem 2.1

In [17]:
# load the dataset csv file into a pandas dataframe
heart_df = pd.read_csv('heart_failure_clinical_records_dataset.csv')


### (a)

I believe that this is a supervised machine learning task. This is because we are given a label called 'DEATH_EVENT' which is described as "If the patient deceased during the follow-up period (boolean)". 

### (b)

In [18]:
print("Axes:", heart_df.axes)
print("-----------------------------------------------------------------------------------")
print("Data types:")
display(heart_df.dtypes)

Axes: [RangeIndex(start=0, stop=299, step=1), Index(['age', 'anaemia', 'creatinine_phosphokinase', 'diabetes',
       'ejection_fraction', 'high_blood_pressure', 'platelets',
       'serum_creatinine', 'serum_sodium', 'sex', 'smoking', 'time',
       'DEATH_EVENT'],
      dtype='object')]
-----------------------------------------------------------------------------------
Data types:


age                         float64
anaemia                       int64
creatinine_phosphokinase      int64
diabetes                      int64
ejection_fraction             int64
high_blood_pressure           int64
platelets                   float64
serum_creatinine            float64
serum_sodium                  int64
sex                           int64
smoking                       int64
time                          int64
DEATH_EVENT                   int64
dtype: object

### (c)



In [19]:
display(heart_df.head(10))

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0  75.0        0                       582         0                 20   
1  55.0        0                      7861         0                 38   
2  65.0        0                       146         0                 20   
3  50.0        1                       111         0                 20   
4  65.0        1                       160         1                 20   
5  90.0        1                        47         0                 40   
6  75.0        1                       246         0                 15   
7  60.0        1                       315         1                 60   
8  65.0        0                       157         0                 65   
9  80.0        1                       123         0                 35   

   high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                    1  265000.00               1.9           130    1   
1                    0  263358.03               1.1           136    1   
2                    0  162000.00               1.3           129    1   
3                    0  210000.00               1.9           137    1   
4                    0  327000.00               2.7           116    0   
5                    1  204000.00               2.1           132    1   
6                    0  127000.00               1.2           137    1   
7                    0  454000.00               1.1           131    1   
8                    0  263358.03               1.5           138    0   
9                    1  388000.00               9.4           133    1   

   smoking  time  DEATH_EVENT  
0        0     4            1  
1        0     6            1  
2        1     7            1  
3        0     7            1  
4        0     8            1  
5        1     8            1  
6        0    10            1  
7        1    10            1  
8        0    10            1  
9        1    10            1

### (d)

In [20]:
print("Rows:", heart_df.shape[0])
print("Columns:", heart_df.shape[1])

Rows: 299
Columns: 13


### (e)

There are 299 samples in this dataset.

### (f)

There are 12 independent variables in this dataset, which are:  
age  
anaemia  
creatinine_phosphokinase  
diabetes  
ejection_fraction  
high_blood_pressure  
platelets  
serum_creatinine  
serum_sodium  
sex  
smoking  
time  


### (g)

There is 1 dependent variable in this dataset which is:  
DEATH_EVENT

### (h)

n = 299  
p = 12

### (i)

n = 299  
m = 1

### (j)

In [21]:
for col in heart_df.columns:
    unique_vals = pd.unique(heart_df[col])
    if len(unique_vals) < 10:
        print(f"Column: {col}, Unique values: {len(unique_vals)} Values: {unique_vals}")


Column: anaemia, Unique values: 2 Values: [0 1]
Column: diabetes, Unique values: 2 Values: [0 1]
Column: high_blood_pressure, Unique values: 2 Values: [1 0]
Column: sex, Unique values: 2 Values: [1 0]
Column: smoking, Unique values: 2 Values: [0 1]
Column: DEATH_EVENT, Unique values: 2 Values: [1 0]


There are a total of 6 categorical variables which are:
anaemia  
diabetes  
high_blood_pressure  
sex  
smoking  
DEATH_EVENT  

These are all categorical variables since they all are basically yes or no questions, ie binary answers either 1 or 0. The only one that is not a yes or no question is 'sex', this is still categorical as they define sex as either woman or man.  

### (k)

They all follow Ordinal binary encoding {0, 1}

### (l)

In [22]:
print('Number of samples that correspond to patients that have died are:', heart_df['DEATH_EVENT'].value_counts()[1])
print('Number of samples that correspond to patients that have not died are:', heart_df['DEATH_EVENT'].value_counts()[0])


Number of samples that correspond to patients that have died are: 96
Number of samples that correspond to patients that have not died are: 203


### (m)

In [23]:
print('The description does not specify which binary input corresponds to which sex; \nhowever after looking at comments under the dataset people are saying that \n1 corresponds to male patients while 0 corresponds to Women patients; \nTherefore using that assumption we get that:')
print('\nNumber of samples that correspond to Women patients are:', heart_df['sex'].value_counts()[0])
print('Number of samples that correspond to Men patients are:', heart_df['sex'].value_counts()[1])

The description does not specify which binary input corresponds to which sex; 
however after looking at comments under the dataset people are saying that 
1 corresponds to male patients while 0 corresponds to Women patients; 
Therefore using that assumption we get that:

Number of samples that correspond to Women patients are: 105
Number of samples that correspond to Men patients are: 194


### (n)

In [24]:
print('Number of Samples that correspond to smokers are:', heart_df['smoking'].value_counts()[1])
print('Number of Samples that correspond to non-smokers are:', heart_df['smoking'].value_counts()[0])


Number of Samples that correspond to smokers are: 96
Number of Samples that correspond to non-smokers are: 203


## Problem 2.2

In [25]:
pd.reset_option('display.max_rows')

heart_df['age'] = np.where(
    (heart_df['age'] < 0) | 
    (heart_df['age'] > 122) | 
    (heart_df['age'].apply(lambda x: (not pd.isnull(x)) and (x != int(x)))), # Need to check if null first because you cast int() on NaN values
    np.nan, heart_df['age']
)


heart_df['anaemia'] = np.where(
    (heart_df['anaemia'] != 0) & 
    (heart_df['anaemia'] != 1),
    np.nan, heart_df['anaemia']
)

heart_df['creatinine_phosphokinase'] = np.where(
    (heart_df['creatinine_phosphokinase'] <= 0) |
    (heart_df['creatinine_phosphokinase'] > 5000),
    np.nan, heart_df['creatinine_phosphokinase']
)

heart_df['diabetes'] = np.where(
    (heart_df['diabetes'] != 0) & 
    (heart_df['diabetes'] != 1),
    np.nan, heart_df['diabetes']
)

heart_df['ejection_fraction'] = np.where(
    (heart_df['ejection_fraction'] < 15) |
    (heart_df['ejection_fraction'] > 75),
    np.nan, heart_df['ejection_fraction']
)

heart_df['high_blood_pressure'] = np.where(
    (heart_df['high_blood_pressure'] != 0) & 
    (heart_df['high_blood_pressure'] != 1),
    np.nan, heart_df['high_blood_pressure']
)

heart_df['platelets'] = np.where(
    (heart_df['platelets'] < 10) | 
    (heart_df['platelets'] > 1000000),
    np.nan, heart_df['platelets']
)

heart_df['serum_creatinine'] = np.where(
    (heart_df['serum_creatinine'] < 0.5) | 
    (heart_df['serum_creatinine'] > 10),
    np.nan, heart_df['serum_creatinine']
)

heart_df['serum_sodium'] = np.where(
    (heart_df['serum_sodium'] < 120) | 
    (heart_df['serum_sodium'] > 155),
    np.nan, heart_df['serum_sodium']
)

heart_df['sex'] = np.where(
    (heart_df['sex'] != 0) & 
    (heart_df['sex'] != 1),
    np.nan, heart_df['sex']
)

heart_df['smoking'] = np.where(
    (heart_df['smoking'] != 0) & 
    (heart_df['smoking'] != 1),
    np.nan, heart_df['smoking']
)

heart_df['time'] = np.where(
    (heart_df['time'] < 0),
    np.nan, heart_df['time']
)

heart_df['smoking'] = np.where(
    (heart_df['smoking'] != 0) & 
    (heart_df['smoking'] != 1),
    np.nan, heart_df['smoking']
)




age: Got rid of negative ages as well as values that are not integers since age should be an integer. Also got rid of ages greater than 122 since the oldest person to live died at 122.  

creatinine_phosphokinase: Got rid of negative values as negative values for this variable does not make sense. Also found that the normal range of this value is from 20 - 2000 so I got rid of values larger than 5000 considering that this dataset contains people who have died.

ejection_fraction: Got rid of values below 15% and above 75% as those values may indicate an error in the equipment.

platelets: Numbers outside of the range of 10 - 100k is most likely equipment error.

serum_creatinine: Numbers outside of the range of 0.5 - 10 is most likely equipment error.

serum_sodium: Numbers outside of the range of 120 - 155 is most likely equipment error.

time: Numbers less than 0 does not make sense for this.

The rest of the variables are binary categorical data, so I replaced values that are not 0 or 1 with NaN.


 



## Problem 2.3

In [26]:

def zero_mean_and_unit_variance(df, cols_to_exclude):
    df.loc[:, df.columns.difference(cols_to_exclude)] = df.loc[:, df.columns.difference(cols_to_exclude)].apply(lambda x: (x - x.mean()) / x.std())

categorical_cols = ['anaemia','diabetes', 'high_blood_pressure', 'sex','smoking', 'DEATH_EVENT']
zero_mean_and_unit_variance(heart_df, categorical_cols)

display(heart_df.head(20))

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0   1.186852      0.0                  0.126530       0.0          -1.562190   
1  -0.488903      0.0                       NaN       0.0          -0.002043   
2   0.348975      0.0                 -0.541160       0.0          -1.562190   
3  -0.907842      1.0                 -0.594759       0.0          -1.562190   
4   0.348975      1.0                 -0.519720       1.0          -1.562190   
5   2.443668      1.0                 -0.692768       0.0           0.171307   
6   1.186852      1.0                 -0.388020       0.0          -1.995565   
7  -0.069964      1.0                 -0.282354       1.0           1.904804   
8   0.348975      0.0                 -0.524315       0.0           2.338178   
9   1.605791      1.0                 -0.576382       0.0          -0.262067   
10  1.186852      1.0                 -0.640701       0.0          -0.002043   
11  0.097611      0.0                 -0.410991       0.0          -1.128816   
12 -1.326780      1.0                  0.737558       0.0          -0.695442   
13 -0.907842      1.0                 -0.507469       0.0          -0.002043   
14 -0.991629      1.0                 -0.642232       0.0          -0.695442   
15  1.773366      1.0                 -0.184344       0.0           1.038055   
16  2.192305      1.0                 -0.536566       0.0          -0.002043   
17 -1.326780      0.0                  0.126530       0.0                NaN   
18  0.767913      1.0                 -0.573319       0.0          -1.128816   
19 -1.075417      1.0                  0.126530       1.0           1.471430   

    high_blood_pressure     platelets  serum_creatinine  serum_sodium  sex  \
0                   1.0  1.678834e-02          0.489237     -1.680195  1.0   
1                   0.0  7.523047e-09         -0.284076     -0.192070  1.0   
2                   0.0 -1.036336e+00         -0.090748     -1.928216  1.0   
3                   0.0 -5.455595e-01          0.489237      0.055951  1.0   
4                   0.0  6.507077e-01          1.262550           NaN  0.0   
5                   1.0 -6.069065e-01          0.682565     -1.184153  1.0   
6                   0.0 -1.394193e+00         -0.187412      0.055951  1.0   
7                   0.0  1.949220e+00         -0.284076     -1.432174  1.0   
8                   0.0  7.523047e-09          0.102580      0.303972  0.0   
9                   1.0  1.274403e+00          7.739045     -0.936133  1.0   
10                  1.0  1.069912e+00          2.519183     -1.432174  1.0   
11                  1.0 -1.059057e-01         -0.477404      0.800013  1.0   
12                  0.0 -1.302173e+00         -0.284076      0.055951  1.0   
13                  1.0  1.292579e-01         -0.284076      0.055951  1.0   
14                  1.0  1.673158e+00         -0.380740      0.303972  0.0   
15                  0.0 -2.212154e+00         -0.090748     -0.192070  1.0   
16                  0.0 -1.388518e-02         -0.477404      0.800013  1.0   
17                  0.0 -9.954377e-01         -0.574068     -2.424258  1.0   
18                  1.0 -2.694978e-01         -0.380740      0.800013  0.0   
19                  0.0 -1.803174e+00          0.489237     -3.912383  0.0   

    smoking      time  DEATH_EVENT  
0       0.0 -1.626775            1  
1       0.0 -1.601007            1  
2       1.0 -1.588122            1  
3       0.0 -1.588122            1  
4       0.0 -1.575238            1  
5       1.0 -1.575238            1  
6       0.0 -1.549470            1  
7       1.0 -1.549470            1  
8       0.0 -1.549470            1  
9       1.0 -1.549470            1  
10      1.0 -1.549470            1  
11      1.0 -1.549470            1  
12      0.0 -1.536586            1  
13      0.0 -1.536586            1  
14      0.0 -1.523701            0  
15      0.0 -1.510817            1  
16      0.0 -1.497933            1  
17      0.0 -1.497933  

## Problem 2.4

In [27]:
one_hot_encoded = pd.get_dummies(heart_df, columns=['DEATH_EVENT'], prefix='DEATH_EVENT')

one_hot_encoded.to_csv('heart_failure_dataset_processed.csv', index=False)

## Problem 2.5

In [28]:
correlation_matrix = heart_df.corr()

death_event_corr = correlation_matrix['DEATH_EVENT'].sort_values(ascending=False)
display(death_event_corr)

DEATH_EVENT                 1.000000
serum_creatinine            0.294278
age                         0.254871
high_blood_pressure         0.079351
anaemia                     0.066270
diabetes                   -0.001943
sex                        -0.004316
smoking                    -0.012623
creatinine_phosphokinase   -0.039709
platelets                  -0.049139
serum_sodium               -0.203677
ejection_fraction          -0.258994
time                       -0.526964
Name: DEATH_EVENT, dtype: float64

### (a)

The variables that are most positively correlated with 'DEATH_EVENT' are 'DEATH_EVENT' itself and 'serum_creatine'.

### (b)

The variables that are most negatively correlated with 'DEATH_EVENT' are 'time' and 'ejection_fraction'.

### (c)

It's logical that serum creatinine has a positive correlation with death because serum creatinine is a marker of kidney function and higher levels of it often indicate a bad kidney which can lead to death.

### (d)

It's logical that time is negatively correlated with the DEATH_EVENT variable because time indicates the period between the initial visit and the follow visit, so the longer this period is, the longer the patient has survived, indicating that their DEATH_EVENT value is 0.

It's also logical that ejection fraction is negatively correlated with the DEATH_EVENT variable because ejection fraction measures the percentage of blood pumped out of the heart with each beat, therefore the lower this number is the higher the probability of the patient dying.